plugins/org.eclipse.elk.core.service/src/org/eclipse/elk/core/service/util/CompoundGraphElementVisitor.java

```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0  
+ *******************************************************************************/
+package org.eclipse.elk.core.service.util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.elk.core.util.ElkUtil;
+import org.eclipse.elk.core.util.IGraphElementVisitor;
+import org.eclipse.elk.graph.ElkGraphElement;
+import org.eclipse.elk.graph.ElkNode;
+
+/**
+ * Class to apply many graph element visitors on the same graph.
+ */
+public class CompoundGraphElementVisitor implements IGraphElementVisitor {
+    
+    /**
+     * List of graph visitors which will be applied in the given order.
+     */
+    private List<IGraphElementVisitor> graphVisitors = new ArrayList<>();
+    
+    /**
+     * True, if the first visitor is applied to the complete graph before the second one is applied.
+     * False, if all visitors are first applied to the first node and after that to the second node... .
+     */
+    private boolean applyToFullGraphFirst = false;
+    
+    /**
+     * Creates a compound graph visitor using the given visitors.
+     * 
+     * @param visitors The graph visitors
+     */
+    public CompoundGraphElementVisitor(final IGraphElementVisitor... visitors) {
+        graphVisitors.addAll(Arrays.asList(visitors));
+    }
+
+    /**
+     * Creates a compound graph visitor using the given visitors.
+     * 
+     * @param applyVisitorsDepthFirst Whether the first visitor is applied to all element first (true) or
+     * all visitors first to the first element (false)
+     * @param visitors The graph visitors
+     */
+    public CompoundGraphElementVisitor(final boolean applyToFullGraphFirst, final IGraphElementVisitor... visitors) {
+        this(visitors);
+        this.applyToFullGraphFirst = applyToFullGraphFirst;
+    }
+
+    /* (non-Javadoc)
+     * @see org.eclipse.elk.core.util.IGraphElementVisitor#visit(org.eclipse.elk.graph.ElkGraphElement)
+     */
+    @Override
+    public void visit(final ElkGraphElement element) {
+        if (element instanceof ElkNode) {
```

Comment:
Why is this only done for nodes? `IGraphElementVisitor`s visit graph elements, not just nodes.

Reply:
`ElkUtil.applyVisitors` only supports ElkNodes. I suggest to drop non default apply option because of this.

Reply:
I'm not sure I see the problem. `ElkUtil.applyVisitors(...)` expects to be called on a node and then applies the visitors to each graph element in that node and its descendants. That is exactly what we want here, but we should only call the method on the root node.

Am I missing something??

Reply:
No, I thought this would result in a different behavior than the default case. I will just implement it to work like this.